### DISCUSSION TITLE
VS 2019 - ClickOnce Publish - Could not load file or assembly "CefSharp.Core.Runtime.dll" or one of its dependencies.

### DISCUSSION THREAD

**Discussion by seltix5** [cefsharp/CefSharp#4023] [General]
- **What version of the product are you using?**
    NuGet:
    cef.redist.x64 - 98.1.21,
    cef.redist.x86 - 98.1.21,
    CefSharp.Common - 98.1.210,
    CefSharp.WinForms - 98.1.210

- **What architecture x86 or x64?**
    x64
	
- **What version of .Net?**
    .Net 4.7.2
    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WinForms
    
- **What steps will reproduce the problem?**
    Publish the project and execute it, it will give the error.
    When running it directly from VS or from the bin folder it works as expected. When executing from the project publish it will give that error despite the fact the referred file was deployed.

![image](https://user-images.githubusercontent.com/11258096/156414787-f0bf6abc-53fe-4550-aae2-5dec50e61837.png)
```

See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.IO.FileNotFoundException: Could not load file or assembly 'CefSharp.Core.Runtime.dll' or one of its dependencies. The specified module could not be found.
File name: 'CefSharp.Core.Runtime.dll'
   at CefSharp.Cef.get_IsInitialized()
   at CefSharp.WinForms.ChromiumWebBrowser.InitializeFieldsAndCefIfRequired()
   at CefSharp.WinForms.ChromiumWebBrowser.OnHandleCreated(EventArgs e)
   at System.Windows.Forms.Control.WmCreate(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)

```

**Comment by amaitland** [comment]
Converting to discussion as this is unlikely a bug. 

Publishing within Visual Studio requires additional steps.

https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-757106286

The minimum example provides a working example. 

https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj#L31

**Comment by seltix5** [comment]
hi. thanks.
I already try that but did not work.
I changed the platform target from "any cpu" to "x64" and updated the csproj as described.
![image](https://user-images.githubusercontent.com/11258096/156467835-81f52128-c09a-4a02-8d95-cf78a782b7db.png)

**Reply by amaitland** [reply]
Please download and test with the minimal example. 

https://github.com/cefsharp/CefSharp.MinimalExample

Please no images of code. Remove the CefSharpAnyCpuSupport property if you are targeting x64. 

Are you testing on a different machine? Did you download and install the visual c++ runtime? https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022

**Reply by amaitland** [reply]
Detailed instructions for use in the context of the minimal example are at https://github.com/cefsharp/CefSharp/discussions/3613#discussioncomment-913061

The change has actually been made already, same steps should apply to your project file. 

Alternatively migrate your csproj from using packages.config to PackageReference see https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference

**Reply by seltix5** [reply]
the minimal example does work but its not ClickOnce publish.

i'm testing on the same machine where i'm developing and where it works if I run it from the bin folder. I do have both c++ runtime installed (x86 & x64).

**Reply by amaitland** [reply]
> the minimal example does work but its not ClickOnce publish.

Publish the minimal example and test it.

**Reply by seltix5** [reply]
I already found your post #3613 and already added CefSharpBuildAction and Microsoft.Windows.Common-Controls was already commented.

I also already converted the packages.config to PackageReference

**Reply by seltix5** [reply]
> Publish the minimal example and test it.

How can I do that? there is no ClickOnce publish option

**Reply by amaitland** [reply]
> I already found your post #3613 and already added CefSharpBuildAction and Microsoft.Windows.Common-Controls was already commented.

In future please detail everything you've tried. I'm wasting my time repeating myself if you've already read the links.

> How can I do that? there is no ClickOnce publish option

Which solution did you open?

**Reply by seltix5** [reply]
> In future please detail everything you've tried. I'm wasting my time repeating myself if you've already read the links.

Sorry, Did not know if it was relevant or not. I already try everything I could find online.

> Which solution did you open?

CefSharp.MinimalExample.WinForms.net472

**Reply by amaitland** [reply]
> CefSharp.MinimalExample.WinForms.net472

You need to open `CefSharp.MinimalExample.sln` for the older style project. The newer SDK style projects don't support publish (Microsoft limitation)

**Reply by amaitland** [reply]
Using the `Windows Sandbox` for testing, installing VC_redist.x64.exe and installing a `x64` build published from `CefSharp.MinimalExample.WinForms` works as expected.

**Reply by amaitland** [reply]
If you are still having problems then upload a [Minimal Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) to `GitHub` (no zip files). Keep your changes to a  minimum and I will have a look.

**Reply by seltix5** [reply]
> You need to open `CefSharp.MinimalExample.sln` for the older style project. The newer SDK style projects don't support publish (Microsoft limitation)

Ok, I open that one and removed the "Microsoft.Windows.Common-Controls" from app.manifest. The normal execution works, but the ClickOnce publish does not start ( no error is shown )

I also noticed in this project the x86 and x64 folders contain multiple files that are not available on my project.

I will try windows sandbox next.

**Reply by seltix5** [reply]
I installed the ClickOnce on the windows sandbox but same result, the application does not open ( probably crashing before opening )

**Reply by amaitland** [reply]
Did you install visual c++ in the sandbox? 

Have you checked the event viewer for errors? 

Did you publish for x86 or x64? AnyCPU (default When opening the project) is not supported.

**Reply by seltix5** [reply]
> Did you publish for x86 or x64? AnyCPU (default When opening the project) is not supported.

ok, changed to x64 and now it works as expected (on my pc). so the problem must be on the publish related settings. after comparing the two I can see not all files are been deployed. I only have:

CefSharp.Core.dll
CefSharp.dll
CefSharp.WinForms.dll

what can I do to fix that?
Thanks!

**Reply by amaitland** [reply]
Converting `CefSharp.MinimalExample.WinForms.csproj` to use `PackageReference` and publishing an `x64` version also worked as expected when deployed to the `Windows Sandbox` (after installing `VC++ x64`).

> what can I do to fix that?

You've provided so little information it's hard for me to speculate. Best guess would be your project file is incorrect.

How does your project file compare to the working minimal example?


> If you are still having problems then upload a [Minimal Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) to `GitHub` (no zip files). Keep your changes to a minimum and I will have a look.

Have you attempted to create a reproducible example?

**Reply by seltix5** [reply]
ok that was it. found the solution.
to my .csproj I added :
 ```
 <Import Project="..\packages\CefSharp.Common.98.1.210\build\CefSharp.Common.props" Condition="Exists('..\packages\CefSharp.Common.98.1.210\build\CefSharp.Common.props')" />
  <Import Project="..\packages\cef.redist.x86.98.1.21\build\cef.redist.x86.props" Condition="Exists('..\packages\cef.redist.x86.98.1.21\build\cef.redist.x86.props')" />
  <Import Project="..\packages\cef.redist.x64.98.1.21\build\cef.redist.x64.props" Condition="Exists('..\packages\cef.redist.x64.98.1.21\build\cef.redist.x64.props')" />
```
like you have on your example https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj
Not sure if that was supposed to be added manually or not but it did solve the problem.
Thanks for the help!

**Reply by amaitland** [reply]
> > like you have on your example https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj

When you converted to `PackageReference` these should have been removed. This will likely break next time you upgrade.

> Not sure if that was supposed to be added manually or not but it did solve the problem.

Only valid for the older `packages.config`. If you successfully upgraded to `PackageReference` then the packages folder should have been removed.

If you plan on using `PackageReference` then I'd suggest upgrading the `MinimalExample` and then comparing it.

**Reply by seltix5** [reply]
I did convert the main project to PackageReference, but not all projects, the solution has around 50, and thus the packages folder still exists.
Now that you talk about it, I converted the main project to PackageReference but not the project where the CefSharp user controls are... Maybe that was my problme... I will test to remove the lines from my main .csproj and convert the project where the user controls are to see if that would also solve the problem and report back

**Reply by amaitland** [reply]
> I did convert the main project to PackageReference, but not all projects, the solution has around 50, and thus the packages folder still exists.

This is really **important** information. Converting your main project by itself wouldn't do anything.

**Reply by seltix5** [reply]
Hi. This is so annoying... I cant reproduce the problem anymore -.- I removed those lines from the main .csproj and rebuild and it was working anyway. I then deleted the bin folder, uninstalled/installed the CefSharp nuget  (again, on the main project + subproject where it is used as, as before) (at this moment I could confirm the `Import` lines was added automatically to the subproject) but its working anyway.... 

Structure (both with nuget CefSharp.WinForms 98.1.210):
main.csproj - PackageReference
subproject.csproj - packages.config ( with the 3 Import lines to the other 3 nuget dependencies, CefSharp.Common, cef.redist.x86 & cef.redist.x64 )

I'm sorry for all the trouble :s I hope at least this discussion may help others in the future.

**Comment by amaitland** [comment]
For anyone else that ends up here I've created a more detailed guide for publishing using `ClickOnce` see https://github.com/cefsharp/CefSharp/discussions/4056
